File: WindowTracker.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;

namespace WheelSwitcher;

/// <summary>
/// Describes a window slotted onto the wheel.
/// </summary>
public sealed class TrackedWindow
{
    public IntPtr Handle { get; init; }
    public string Title { get; set; } = "";
    public string ProcessPath { get; set; } = "";
    public int Slot { get; set; } // 0..7
    public DateTime LastSeen { get; set; }

    public override string ToString() => $"[{Slot}] {Title}";
}

public sealed class WindowTracker
{
    public const int MaxSlots = 8;

    // Slot -> Window (or null if empty)
    private readonly TrackedWindow?[] _slots = new TrackedWindow?[MaxSlots];

    // Handle -> slot index, so we keep assignments stable across refreshes.
    private readonly Dictionary<IntPtr, int> _handleToSlot = new();

    // Windows that didn't fit in the wheel (Z-order/MRU sorted).
    private readonly List<TrackedWindow> _overflow = new();

    private readonly IntPtr _selfWindow;

    public WindowTracker(IntPtr selfWindow)
    {
        _selfWindow = selfWindow;
    }

    public IReadOnlyList<TrackedWindow?> Slots => _slots;
    public IReadOnlyList<TrackedWindow> Overflow => _overflow;

    // The window to pre-select when the wheel opens: second in Z-order (MRU), which is
    // the window that was active before the current foreground. EnumWindows is Z-order.
    public IntPtr PreviousWindowHandle { get; private set; }

    /// <summary>
    /// Enumerates current top-level windows, prunes closed ones, and fills empty slots
    /// in MRU-ish (Z-order) order up to MaxSlots.
    /// </summary>
    public void Refresh()
    {
        var live = EnumerateAltTabWindows();

        // live is Z-order: live[0] = current foreground, live[1] = previously active window.
        PreviousWindowHandle = live.Count > 1 ? live[1].Handle : IntPtr.Zero;

        // 1. Drop slots whose handle is gone.
        for (int i = 0; i < MaxSlots; i++)
        {
            var s = _slots[i];
            if (s is null)
                continue;
            if (!live.Any(w => w.Handle == s.Handle))
            {
                _handleToSlot.Remove(s.Handle);
                WindowActivator.EvictIcon(s.Handle);
                _slots[i] = null;
            }
        }

        // 2. Update titles for surviving windows.
        foreach (var w in live)
        {
            if (_handleToSlot.TryGetValue(w.Handle, out int slot))
            {
                var tracked = _slots[slot]!;
                tracked.Title = w.Title;
                tracked.LastSeen = DateTime.UtcNow;
            }
        }

        // 3. Add new windows into the lowest free slot; overflow the rest. Z-order is MRU-first.
        _overflow.Clear();
        foreach (var w in live)
        {
            if (_handleToSlot.ContainsKey(w.Handle))
                continue;

            int free = FindFreeSlot();
            if (free >= 0)
            {
                w.Slot = free;
                _slots[free] = w;
                _handleToSlot[w.Handle] = free;
            }
            else
            {
                _overflow.Add(w);
            }
        }
    }

    private int FindFreeSlot()
    {
        for (int i = 0; i < MaxSlots; i++)
            if (_slots[i] is null)
                return i;
        return -1;
    }

    /// <summary>Swap two slot contents (used for drag-to-reorder).</summary>
    public void SwapSlots(int a, int b)
    {
        if (a == b)
            return;
        if (a < 0 || a >= MaxSlots || b < 0 || b >= MaxSlots)
            return;

        (_slots[a], _slots[b]) = (_slots[b], _slots[a]);
        if (_slots[a] is not null)
        {
            _slots[a]!.Slot = a;
            _handleToSlot[_slots[a]!.Handle] = a;
        }
        if (_slots[b] is not null)
        {
            _slots[b]!.Slot = b;
            _handleToSlot[_slots[b]!.Handle] = b;
        }
    }

    /// <summary>Swap a wheel slot with an overflow entry.</summary>
    public void SwapSlotWithOverflow(int slotIndex, int overflowIndex)
    {
        if (slotIndex < 0 || slotIndex >= MaxSlots)
            return;
        if (overflowIndex < 0 || overflowIndex >= _overflow.Count)
            return;

        var slotWindow = _slots[slotIndex];
        var overflowWindow = _overflow[overflowIndex];

        overflowWindow.Slot = slotIndex;
        _slots[slotIndex] = overflowWindow;
        _handleToSlot[overflowWindow.Handle] = slotIndex;

        if (slotWindow is not null)
        {
            _handleToSlot.Remove(slotWindow.Handle);
            slotWindow.Slot = -1;
            _overflow[overflowIndex] = slotWindow;
        }
        else
        {
            _overflow.RemoveAt(overflowIndex);
        }
    }

    /// <summary>Reorder two overflow entries.</summary>
    public void SwapOverflowItems(int a, int b)
    {
        if (a == b)
            return;
        if (a < 0 || a >= _overflow.Count || b < 0 || b >= _overflow.Count)
            return;
        (_overflow[a], _overflow[b]) = (_overflow[b], _overflow[a]);
    }

    // ---- Test helpers ----
    internal void ForceSlot(int index, TrackedWindow? window)
    {
        if (_slots[index] is TrackedWindow old)
            _handleToSlot.Remove(old.Handle);
        _slots[index] = window;
        if (window is not null)
        {
            window.Slot = index;
            _handleToSlot[window.Handle] = index;
        }
    }

    internal void ForceOverflow(TrackedWindow window) => _overflow.Add(window);

    // --- Enumeration ---
    private List<TrackedWindow> EnumerateAltTabWindows()
    {
        var result = new List<TrackedWindow>();
        IntPtr shell = NativeMethods.GetShellWindow();

        NativeMethods.EnumWindows(
            (hWnd, _) =>
            {
                if (hWnd == shell)
                    return true;
                if (hWnd == _selfWindow)
                    return true;
                if (!IsAltTabEligible(hWnd))
                    return true;

                var title = GetWindowTitle(hWnd);
                if (string.IsNullOrWhiteSpace(title))
                    return true;

                result.Add(
                    new TrackedWindow
                    {
                        Handle = hWnd,
                        Title = title,
                        ProcessPath = GetProcessPath(hWnd),
                        LastSeen = DateTime.UtcNow,
                    }
                );
                return true;
            },
            IntPtr.Zero
        );

        return result;
    }

    /// <summary>
    /// Matches (roughly) the set of windows the real Alt+Tab switcher shows.
    /// Key filters: visible, not cloaked (UWP trick), not a tool window unless app-window,
    /// and - for UWP - skip the invisible ApplicationFrameHost wrapper duplicates.
    /// </summary>
    private static bool IsAltTabEligible(IntPtr hWnd)
    {
        if (!NativeMethods.IsWindowVisible(hWnd))
            return false;

        // Skip owned windows (like dialogs whose owner is already listed)... actually we DO want popups,
        // but only if they are top-level. Matching real Alt+Tab: show windows where the root owner is itself.
        IntPtr owner = NativeMethods.GetWindow(hWnd, NativeMethods.GW_OWNER);
        int exStyle = NativeMethods.GetWindowLong(hWnd, NativeMethods.GWL_EXSTYLE);

        bool isTool = (exStyle & NativeMethods.WS_EX_TOOLWINDOW) != 0;
        bool isAppWindow = (exStyle & NativeMethods.WS_EX_APPWINDOW) != 0;

        if (owner != IntPtr.Zero && !isAppWindow)
            return false;
        if (isTool && !isAppWindow)
            return false;

        // Cloaked check - suspended UWP apps and apps on other virtual desktops report cloaked != 0.
        if (
            NativeMethods.DwmGetWindowAttribute(
                hWnd,
                NativeMethods.DWMWA_CLOAKED,
                out int cloaked,
                sizeof(int)
            ) == 0
            && cloaked != 0
        )
            return false;

        // Filter out empty-class shell helpers.
        var cls = new StringBuilder(256);
        NativeMethods.GetClassName(hWnd, cls, cls.Capacity);
        string className = cls.ToString();
        if (
            className is "Windows.UI.Core.CoreWindow" or "ApplicationFrameWindow"
            && NativeMethods.IsIconic(hWnd)
        )
        {
            // Minimised UWP windows are fine; it's the invisible duplicates we drop via cloaked check above.
        }

        return true;
    }

    private static string GetWindowTitle(IntPtr hWnd)
    {
        int len = NativeMethods.GetWindowTextLength(hWnd);
        if (len <= 0)
            return "";
        var sb = new StringBuilder(len + 1);
        NativeMethods.GetWindowText(hWnd, sb, sb.Capacity);
        return sb.ToString();
    }

    private static string GetProcessPath(IntPtr hWnd)
    {
        NativeMethods.GetWindowThreadProcessId(hWnd, out uint pid);
        if (pid == 0)
            return "";
        IntPtr h = NativeMethods.OpenProcess(
            NativeMethods.PROCESS_QUERY_LIMITED_INFORMATION,
            false,
            pid
        );
        if (h == IntPtr.Zero)
            return "";
        try
        {
            var sb = new StringBuilder(1024);
            int cap = sb.Capacity;
            if (NativeMethods.QueryFullProcessImageName(h, 0, sb, ref cap))
                return sb.ToString();
        }
        finally
        {
            NativeMethods.CloseHandle(h);
        }
        return "";
    }
}

File: src/window_tracker.rs
use std::collections::HashMap;
use windows::Win32::Foundation::{BOOL, HWND, LPARAM};
use windows::Win32::Graphics::Dwm::{DwmGetWindowAttribute, DWMWA_CLOAKED};
use windows::Win32::UI::WindowsAndMessaging::{
    EnumWindows, GetShellWindow, GetWindow, GetWindowLongW, GetWindowTextLengthW,
    GetWindowTextW, IsWindowVisible, GW_OWNER, GWL_EXSTYLE, WS_EX_APPWINDOW, WS_EX_TOOLWINDOW,
};

pub const MAX_SLOTS: usize = 8;

#[derive(Clone, Debug)]
pub struct TrackedWindow {
    pub hwnd: HWND,
    pub title: String,
    pub slot: usize,
}

pub struct WindowTracker {
    slots: [Option<TrackedWindow>; MAX_SLOTS],
    // HashMap keyed by HWND as usize (HWND.0 is *mut c_void in windows-rs 0.58)
    handle_to_slot: HashMap<usize, usize>,
    overflow: Vec<TrackedWindow>,
    self_hwnd: HWND,
    pub previous_hwnd: HWND,
}

impl WindowTracker {
    pub fn new(self_hwnd: HWND) -> Self {
        Self {
            slots: Default::default(),
            handle_to_slot: HashMap::new(),
            overflow: Vec::new(),
            self_hwnd,
            previous_hwnd: HWND(std::ptr::null_mut()),
        }
    }

    pub fn slots(&self) -> &[Option<TrackedWindow>; MAX_SLOTS] {
        &self.slots
    }

    pub fn overflow(&self) -> &[TrackedWindow] {
        &self.overflow
    }

    /// Enumerate current alt-tab windows, prune closed ones, fill empty slots.
    pub fn refresh(&mut self) {
        let live = self.enumerate_alt_tab_windows();

        // Z-order: live[0] = current foreground, live[1] = previously active.
        self.previous_hwnd = live.get(1).map(|w| w.hwnd).unwrap_or(HWND(std::ptr::null_mut()));

        // Drop slots whose window is no longer alive.
        for i in 0..MAX_SLOTS {
            if let Some(ref w) = self.slots[i] {
                if !live.iter().any(|lw| lw.hwnd == w.hwnd) {
                    self.handle_to_slot.remove(&(w.hwnd.0 as usize));
                    self.slots[i] = None;
                }
            }
        }

        // Update titles for windows that survived.
        for lw in &live {
            if let Some(&slot) = self.handle_to_slot.get(&(lw.hwnd.0 as usize)) {
                if let Some(ref mut tw) = self.slots[slot] {
                    tw.title = lw.title.clone();
                }
            }
        }

        // Assign new windows to free slots; overflow the rest.
        self.overflow.clear();
        for lw in live {
            if self.handle_to_slot.contains_key(&(lw.hwnd.0 as usize)) {
                continue;
            }
            match self.find_free_slot() {
                Some(slot) => {
                    self.handle_to_slot.insert(lw.hwnd.0 as usize, slot);
                    self.slots[slot] = Some(TrackedWindow { slot, ..lw });
                }
                None => {
                    self.overflow.push(lw);
                }
            }
        }
    }

    fn find_free_slot(&self) -> Option<usize> {
        (0..MAX_SLOTS).find(|&i| self.slots[i].is_none())
    }

    /// Swap two wheel slots.
    pub fn swap_slots(&mut self, a: usize, b: usize) {
        if a == b || a >= MAX_SLOTS || b >= MAX_SLOTS {
            return;
        }
        self.slots.swap(a, b);
        if let Some(ref mut w) = self.slots[a] {
            w.slot = a;
            self.handle_to_slot.insert(w.hwnd.0 as usize, a);
        }
        if let Some(ref mut w) = self.slots[b] {
            w.slot = b;
            self.handle_to_slot.insert(w.hwnd.0 as usize, b);
        }
    }

    /// Move an overflow entry onto a wheel slot (evicting the current occupant to overflow).
    pub fn swap_slot_with_overflow(&mut self, slot: usize, overflow_idx: usize) {
        if slot >= MAX_SLOTS || overflow_idx >= self.overflow.len() {
            return;
        }
        let mut ov_win = self.overflow.remove(overflow_idx);
        ov_win.slot = slot;
        self.handle_to_slot.insert(ov_win.hwnd.0 as usize, slot);

        if let Some(mut evicted) = self.slots[slot].take() {
            self.handle_to_slot.remove(&(evicted.hwnd.0 as usize));
            evicted.slot = usize::MAX;
            self.overflow.insert(overflow_idx, evicted);
        }
        self.slots[slot] = Some(ov_win);
    }

    /// Reorder two overflow entries.
    pub fn swap_overflow(&mut self, a: usize, b: usize) {
        if a != b && a < self.overflow.len() && b < self.overflow.len() {
            self.overflow.swap(a, b);
        }
    }

    // ---- Enumeration ----

    fn enumerate_alt_tab_windows(&self) -> Vec<TrackedWindow> {
        struct State {
            result: Vec<TrackedWindow>,
            shell: HWND,
            self_hwnd: HWND,
        }

        unsafe extern "system" fn enum_proc(hwnd: HWND, lparam: LPARAM) -> BOOL {
            let state = &mut *(lparam.0 as *mut State);
            if hwnd == state.shell || hwnd == state.self_hwnd {
                return BOOL(1);
            }
            if !is_alt_tab_eligible(hwnd) {
                return BOOL(1);
            }
            let title = get_window_title(hwnd);
            if title.is_empty() {
                return BOOL(1);
            }
            state.result.push(TrackedWindow { hwnd, title, slot: 0 });
            BOOL(1)
        }

        let mut state = State {
            result: Vec::new(),
            shell: unsafe { GetShellWindow() },
            self_hwnd: self.self_hwnd,
        };
        unsafe {
            let _ = EnumWindows(Some(enum_proc), LPARAM(&mut state as *mut _ as isize));
        }
        state.result
    }
}

fn is_alt_tab_eligible(hwnd: HWND) -> bool {
    unsafe {
        if !IsWindowVisible(hwnd).as_bool() {
            return false;
        }

        let owner = GetWindow(hwnd, GW_OWNER).unwrap_or_default();
        let ex_style = GetWindowLongW(hwnd, GWL_EXSTYLE) as u32;
        let is_tool = (ex_style & WS_EX_TOOLWINDOW.0) != 0;
        let is_app = (ex_style & WS_EX_APPWINDOW.0) != 0;

        if !owner.0.is_null() && !is_app {
            return false;
        }
        if is_tool && !is_app {
            return false;
        }

        // Cloaked check: suspended UWP apps and windows on other virtual desktops.
        let mut cloaked: u32 = 0;
        if DwmGetWindowAttribute(
            hwnd,
            DWMWA_CLOAKED,
            &mut cloaked as *mut u32 as *mut std::ffi::c_void,
            std::mem::size_of::<u32>() as u32,
        )
        .is_ok()
            && cloaked != 0
        {
            return false;
        }

        true
    }
}

fn get_window_title(hwnd: HWND) -> String {
    unsafe {
        let len = GetWindowTextLengthW(hwnd);
        if len <= 0 {
            return String::new();
        }
        let mut buf = vec![0u16; (len + 1) as usize];
        let copied = GetWindowTextW(hwnd, &mut buf);
        if copied == 0 {
            return String::new();
        }
        String::from_utf16_lossy(&buf[..copied as usize])
    }
}
